### DISCUSSION TITLE
Cannot resolve "fs/promises" error failing build

### DISCUSSION THREAD

**Discussion by Mitsunee** [vercel/next.js#34992] [Help]
### Verify canary release

- [X] I verified that the issue exists in Next.js canary release

### Provide environment information

    Operating System:
      Platform: linux
      Arch: x64
      Version: #33-Ubuntu SMP Fri Feb 4 17:03:31 UTC 2022
    Binaries:
      Node: 14.19.0
      npm: 8.5.2
      Yarn: 1.22.17
      pnpm: 6.32.2
    Relevant packages:
      next: 12.1.0 (also tested 12.1.1-canary.6)
      react: 17.0.2
      react-dom: 17.0.2

### What browser are you using? (if relevant)

not relevant

### How are you deploying your application? (if relevant)

Vercel

### Describe the Bug

Since upgrading to Next.js 12.1.0 running `yarn build` produces this error for any import from `"fs/promises"` including pages, their imports and npm packages:

```
yarn run v1.22.17
$ node --version && next build
v14.19.0
info  - Checking validity of types  
info  - Creating an optimized production build  
Failed to compile.

./src/utils/server/getFileList.js
Module not found: Can't resolve 'fs/promises' in '/home/mitsu/git/Projects/fgo-tools/src/utils/server'
```

This bug was particularly hard to pinpoint as these errors look exactly like the ones you would see when accidentally using Node.js 12 instead of 14 and only appear for `next build` and NOT `next dev`. 

I at first suspected Vercel to be using Node.js 12 and after asking on Twitter added `"engines": {"node":"^14"}` to my package.json (which just produced a warning about ^14 not matching 14.x, but oh well) and changed the build command to `node --version && next build` as seen above. This confirmed that the Vercel deployment was indeed using the exact same version as I was using locally, ruling out Vercel as the issue.

### Expected Behavior

The build succeeds as it did with Next.js 12.0.x and earlier versions.

### To Reproduce

- create a new Next.js project using Next.js version 12.1.0 or canary
- add `import { readFile } from "fs/promises";` to `./pages/index.js`
- run the `next build` command

You should get the following output:

```
$ next build
info  - Checking validity of types  
info  - Creating an optimized production build  
Failed to compile.

./pages/index.js
Module not found: Can't resolve 'fs/promises' in '/home/mitsu/git/Temp/reproduce-fs-import-crash/pages'


> Build failed because of webpack errors
```

--------

Edit: Wanted to clarify more clearly that this issue only appears on 12.1.0 and newer, not older versions. The main branch of my project is still on 12.0.x and deploys the same code with no issues.

**Comment by balazsorban44** [comment]
This is likely because you are either not using the import or using it in the page component. 

Next.js will try to eliminate imports from the client bundle that is used in data fetching methods like `getStaticProps` etc.

Otherwise, it will fail the build as `fs` is not an API in the browser. Make sure you use the import, and that it's only used in the data fetching methods.

You can test this on this page: https://next-code-elimination.vercel.app/

This is described on each data fetching method, example: https://nextjs.org/docs/basic-features/data-fetching/get-static-props

**Reply by Mitsunee** [reply]
I am not using the fs imports in the client. This code is known to build with Next 12.0.x as I said above. I have also provided reproduction steps that reproduce this behaviour without editing the clientside code that comes shipped by create-next-app.

**Reply by Mitsunee** [reply]
Also turning my Issue into a community discussion without even giving me a chance to react is pretty rude. I am not *discussing* something, I am reporting a bug and used the Bug Report template.

**Reply by icyJoseph** [reply]
Hi,

As said above, not using the import might lead to this problem. If you use this tool, https://next-code-elimination.vercel.app/s/RPpgVwzQr, you can see that the import is not removed from the client if not used within one of the server side only functions. 

That's why the reproduction example fails. We need to know more about how the import tree looks like to tackle this.

**Reply by icyJoseph** [reply]
I can't really reproduce this, other than the expected failure.

For example I've made `lib/tools` and there I have:

```js
import { readFile } from "fs/promises";

export const fileReader = (str) => readFile(str);

export const noop = () => {};
```

Either import, `fileReader`, and `noop`, in `pages/index.js` breaks the build, if it is not used within the special server side functions (GSSP, GSP). I've tested this from 12.1.0 down to 12.0.4, and it is consistent and expected.

As said before, I think we need some sketch out of the dependencies on your project. I am also about to jump to 12.1.0 on a very large project, and would like to avoid any surprises.

**Reply by Mitsunee** [reply]
I added this to my reproduction attempt and it does indeed let the build pass:

```js
export async function getStaticProps() {
  readFile;
  return { props: {} }
}
```

Seems to me like I am expecting it to be more strict and eliminate (from the client bundle) all not used imports. It's possible that I forgot to comment out some imports in my `getStaticProps` by accident and that is how I'm now getting "fs/promises" imports in the client bundle. It did remove references to an old cache dir that I'm trying to move to a place that vercel can cache for future builds.

Assuming the above is true it might be a feature idea to be able to configure code elimination to set from which bundle it eliminates imports. In my opinion it should of course check both and entirely remove imports not used in either bundle, but that may possibly break imports like `import "some-package/setup";` that don't actually export anything.

**Reply by icyJoseph** [reply]
Right, I think you can do that in the `webpack` config, but I am not sure about what the implications are:

```js
// next.config.js
/** @type {import('next').NextConfig} */
const nextConfig = {
  reactStrictMode: true,
  webpack(config, { isServer }) {
    if (!isServer) {
      config.resolve.fallback = {
        ...config.resolve.fallback,
        fs: false
      };
    }

    return config;
  }
};

module.exports = nextConfig;

```

And now, even if the import is not pruned, as long as it is not used on client side code, you'll be fine.

**Comment by rimzzlabs** [comment]
I have the same issue but it has different case from this one,

so previously I have 2 axios instance, one for the server side, and one for the client side.

each instance is on separate file, and everything is fine, now I'm trying to move those instance, and re export as named export in one single file like this one:

```tsx
import Axios from 'axios'

const UMAMI_URL = process.env.UMAMI_URL
const SITE_URL = process.env.NEXT_PUBLIC_SITE_URL || 'https://rizkicitra.dev'

const headers = { 'Content-Type': 'application/json' }

export const umamiServer = Axios.create({
  baseURL: UMAMI_URL,
  headers
})

export const umamiClient = Axios.create({
  baseURL: process.env.NODE_ENV === 'development' ? 'http://localhost:3000' : SITE_URL,
  headers
})
```

but now I'm facing this issue now, the app can not resolve any module that runs on server, in my case, it's `readFile` from `fs/promises`, so I have a function that will run only on `getStaticProps`

```ts
import { LOCATION_DIR } from '@/services'

import { readFile } from 'fs/promises'
import matter from 'gray-matter'
import { join } from 'path'

type GetContentBySlug<T> = { content: string; header: { slug: string } & T }

export const getContentBySlug = async <T>(
  /** the path to the directory folder, example: `/blog`
   * **NOTE!** that the slash on the string is required!
   */
  path: string,
  /**
   * read as file name
   */
  slug: string
): Promise<GetContentBySlug<T>> => {
  // read path to file src/data/{path}/[{slug}].mdx
  const dir = join(`${LOCATION_DIR}/${path}`, `${slug}.mdx`)
  // read file with promise based
  const file = await readFile(dir, 'utf8')

  // parse file content with gray matter, extract it's header and content
  const { content, data } = matter(file)

  return {
    header: { ...(data as T), slug },
    content
  }
}
```
this snippet is common for retrieving blog post from `.mdx` file

so what is the problem here? are named export is the problem?
